src/librustc_mir/transform/check_unsafety.rs

```diff
@@ -641,13 +641,17 @@ pub fn check_unsafety(tcx: TyCtxt<'_>, def_id: DefId) {
         }
     }
 
-    let mut unsafe_blocks: Vec<_> = unsafe_blocks.iter().collect();
-    unsafe_blocks.sort_by_cached_key(|(hir_id, _)| tcx.hir().hir_id_to_node_id(*hir_id));
-    let used_unsafe: FxHashSet<_> =
-        unsafe_blocks.iter().flat_map(|&&(id, used)| used.then_some(id)).collect();
-    for &(block_id, is_used) in unsafe_blocks {
-        if !is_used {
-            report_unused_unsafe(tcx, &used_unsafe, block_id);
+    let (mut unsafe_used, mut unsafe_unused): (FxHashSet<_>, Vec<_>) = Default::default();
+    for &(block_id, is_used) in unsafe_blocks.iter() {
+        if is_used {
+            unsafe_used.insert(block_id);
+        } else {
+            unsafe_unused.push(block_id);
         }
     }
+    unsafe_unused.sort_by_cached_key(|hir_id| tcx.hir().span(*hir_id));
```

Comment:
Out of curiosity, why do we have to sort these at all? I guess `unsafe_blocks` is not in source order?

Reply:
Yes, it doesn't appear to be; without the sort two (if I recall correctly) tests with nested `unsafe` blocks fail due to unaligned error messages.

Reply:
Would you mind leaving a comment to this effect? It's not immediately obvious why we need to do this otherwise.

Reply:
Sure, let me just run the tests to include their names.

Reply:
Done :+1: .